src/AudioDeviceTestWidget/AudioDevice/AudioDevice.test.tsx

```diff
@@ -83,4 +83,36 @@ describe('the AudioDevice component', () => {
       expect(wrapper.find(Select).at(0).text()).toEqual('deviceoutput1');
     });
   });
+
+  describe('props.onDeviceChange', () => {
+    let onDeviceChange: () => any;
+
+    beforeEach(() => {
+      onDeviceChange = jest.fn();
+    });
+
+    it('should trigger onDeviceChange when a new device is added', () => {
+      mount(<AudioDevice disabled={false} level={1} kind="audioinput" onDeviceChange={onDeviceChange} />);
+      mockDevices.push({
```

Comment:
Looks like this `push` call (and the `pop` call in the next text) can be removed and these tests still pass. I think that the `expect(onDeviceChange).toHaveBeenCalled();` assertion is detecting the initial call to `onDeviceChange` that is being made when the `useEffect` hook is run on component mount. 

I tried to modify these tests so that `onDeviceChange` is called when the device list is updated, but it turns out that the AudioDevice component is not set up to do that. When the device list changes, `updateSelectedDevice(selectedDevice || devices[0].deviceId);` is called, which sets the selected device to the device that is already selected. Because of the `if (value !== selectedDevice)` check in the `updateSelectedDevice` function, `onDeviceChange` is never actually called when the device list updates. 

Because of all of this, I think it's probably fine just to have one test here:

```
    it('should trigger onDeviceChange when devices are present', () => {
      mount(<AudioDevice disabled={false} level={1} kind="audioinput" onDeviceChange={onDeviceChange} />);
      expect(onDeviceChange).toHaveBeenCalled();
    });
```

Reply:
Thanks for looking more into it! I updated the test.